### ISSUE TITLE
Too many open files

### ISSUE THREAD

**Issue by ntoskrnl** [DataDog/dd-sdk-android#1039]
**Describe what happened**

When we started using DD Android, the number of crashes due to out of memory / too many open files significantly increases.

When I run the app with Datadog logging enabled, I see steady increase of open file descriptors of `cache/dd-log-*` with `(deleted)` status. In our use case, the app has to run in background for extended period of time (days), and on some devices the max open file limit is hit within hours. Eventually the app just crashes. 

I tested it on different android versions with latest Datadog Android SDK and used `lsof -p <pid>` command to check opened files.

**Describe what you expected:**

I expect the number of open file handles doesn't increase over time.

**Additional context**
 - Android OS version: 8.0 or 11 
 - Device Model: Emulator, Google Pixel 2 XL, etc.
 - Datadog SDK version: 1.13.0
 - Versions of any other relevant dependencies (OkHttp, …): OkHttp 4.10.0
 - Proguard configuration: doesn't matter

```
lsof -p 12391 | grep "dd-log"

PID       USER   FD      TYPE             DEVICE  SIZE/OFF       NODE NAME
12391    u0_a120   64w      REG              253,0     27916      33607 /data/data/com.example.android.dev/cache/dd-logs-v2/1663001677320 (deleted)
12391    u0_a120   65w      REG              253,0     27916      33607 /data/data/com.example.android.dev/cache/dd-logs-v2/1663001677320 (deleted)
12391    u0_a120   70w      REG              253,0     27916      33607 /data/data/com.example.android.dev/cache/dd-logs-v2/1663001677320 (deleted)
12391    u0_a120   71w      REG              253,0     27916      33607 /data/data/com.example.android.dev/cache/dd-logs-v2/1663001677320 (deleted)
12391    u0_a120   73w      REG              253,0     27916      33607 /data/data/com.example.android.dev/cache/dd-logs-v2/1663001677320 (deleted)
12391    u0_a120   74w      REG              253,0     27916      33607 /data/data/com.example.android.dev/cache/dd-logs-v2/1663001677320 (deleted)
12391    u0_a120   76w      REG              253,0     27916      33607 /data/data/com.example.android.dev/cache/dd-logs-v2/1663001677320 (deleted)
12391    u0_a120   90w      REG              253,0     27916      33607 /data/data/com.example.android.dev/cache/dd-logs-v2/1663001677320 (deleted)
12391    u0_a120   92w      REG              253,0     27916      33607 /data/data/com.example.android.dev/cache/dd-logs-v2/1663001677320 (deleted)
12391    u0_a120   93w      REG              253,0     28652      33619 /data/data/com.example.android.dev/cache/dd-logs-v2/1663001682074 (deleted)
12391    u0_a120   94w      REG              253,0     28652      33619 /data/data/com.example.android.dev/cache/dd-logs-v2/1663001682074 (deleted)
12391    u0_a120   97w      REG              253,0     27916      33607 /data/data/com.example.android.dev/cache/dd-logs-v2/1663001677320 (deleted)
12391    u0_a120   99w      REG              253,0     27916      33607 /data/data/com.example.android.dev/cache/dd-logs-v2/1663001677320 (deleted)
12391    u0_a120  102w      REG              253,0     27916      33607 /data/data/com.example.android.dev/cache/dd-logs-v2/1663001677320 (deleted)
12391    u0_a120  104w      REG              253,0     27916      33607 /data/data/com.example.android.dev/cache/dd-logs-v2/1663001677320 (deleted)
12391    u0_a120  105w      REG              253,0     27916      33607 /data/data/com.example.android.dev/cache/dd-logs-v2/1663001677320 (deleted)
12391    u0_a120  107w      REG              253,0     27916      33607 /data/data/com.example.android.dev/cache/dd-logs-v2/1663001677320 (deleted)
12391    u0_a120  109w      REG              253,0     27916      33607 /data/data/com.example.android.dev/cache/dd-logs-v2/1663001677320 (deleted)
12391    u0_a120  110w      REG              253,0     27916      33607 /data/data/com.example.android.dev/cache/dd-logs-v2/1663001677320 (deleted)
12391    u0_a120  111w      REG              253,0     27916      33607 /data/data/com.example.android.dev/cache/dd-logs-v2/1663001677320 (deleted)
12391    u0_a120  112w      REG              253,0     27916      33607 /data/data/com.example.android.dev/cache/dd-logs-v2/1663001677320 (deleted)
12391    u0_a120  122w      REG              253,0     27916      33607 /data/data/com.example.android.dev/cache/dd-logs-v2/1663001677320 (deleted)
12391    u0_a120  123w      REG              253,0     27916      33607 /data/data/com.example.android.dev/cache/dd-logs-v2/1663001677320 (deleted)
12391    u0_a120  128w      REG              253,0     27916      33607 /data/data/com.example.android.dev/cache/dd-logs-v2/1663001677320 (deleted)
12391    u0_a120  132w      REG              253,0     27916      33607 /data/data/com.example.android.dev/cache/dd-logs-v2/1663001677320 (deleted)
12391    u0_a120  136w      REG              253,0     27916      33607 /data/data/com.example.android.dev/cache/dd-logs-v2/1663001677320 (deleted)
12391    u0_a120  137w      REG              253,0     27916      33607 /data/data/com.example.android.dev/cache/dd-logs-v2/1663001677320 (deleted)
12391    u0_a120  139w      REG              253,0     27916      33607 /data/data/com.example.android.dev/cache/dd-logs-v2/1663001677320 (deleted)
12391    u0_a120  142w      REG              253,0     27916      33607 /data/data/com.example.android.dev/cache/dd-logs-v2/1663001677320 (deleted)
12391    u0_a120  143w      REG              253,0     27916      33607 /data/data/com.example.android.dev/cache/dd-logs-v2/1663001677320 (deleted)
12391    u0_a120  155r      REG              253,0     24277      33210 /data/data/com.example.android.dev/cache/dd-logs-v2/1662999947924 (deleted)
12391    u0_a120  158w      REG              253,0     27916      33607 /data/data/com.example.android.dev/cache/dd-logs-v2/1663001677320 (deleted)
12391    u0_a120  161w      REG              253,0     28716      33632 /data/data/com.example.android.dev/cache/dd-logs-v2/1663001686837 (deleted)
12391    u0_a120  163w      REG              253,0     28652      33619 /data/data/com.example.android.dev/cache/dd-logs-v2/1663001682074 (deleted)
12391    u0_a120  165w      REG              253,0     27916      33607 /data/data/com.example.android.dev/cache/dd-logs-v2/1663001677320 (deleted)
12391    u0_a120  166w      REG              253,0     27916      33607 /data/data/com.example.android.dev/cache/dd-logs-v2/1663001677320 (deleted)
12391    u0_a120  168w      REG              253,0     27916      33607 /data/data/com.example.android.dev/cache/dd-logs-v2/1663001677320 (deleted)
12391    u0_a120  169w      REG              253,0     27916      33607 /data/data/com.example.android.dev/cache/dd-logs-v2/1663001677320 (deleted)
12391    u0_a120  172w      REG              253,0     27916      33607 /data/data/com.example.android.dev/cache/dd-logs-v2/1663001677320 (deleted)
12391    u0_a120  173w      REG              253,0     27916      33607 /data/data/com.example.android.dev/cache/dd-logs-v2/1663001677320 (deleted)
12391    u0_a120  175w      REG              253,0     27916      33607 /data/data/com.example.android.dev/cache/dd-logs-v2/1663001677320 (deleted)
12391    u0_a120  176w      REG              253,0     27916      33607 /data/data/com.example.android.dev/cache/dd-logs-v2/1663001677320 (deleted)
12391    u0_a120  177w      REG              253,0     27916      33607 /data/data/com.example.android.dev/cache/dd-logs-v2/1663001677320 (deleted)
12391    u0_a120  178w      REG              253,0     27916      33607 /data/data/com.example.android.dev/cache/dd-logs-v2/1663001677320 (deleted)
12391    u0_a120  179w      REG              253,0     28652      33619 /data/data/com.example.android.dev/cache/dd-logs-v2/1663001682074 (deleted)
12391    u0_a120  182w      REG              253,0     28652      33619 /data/data/com.example.android.dev/cache/dd-logs-v2/1663001682074 (deleted)
12391    u0_a120  183w      REG              253,0     28652      33619 /data/data/com.example.android.dev/cache/dd-logs-v2/1663001682074 (deleted)
12391    u0_a120  184w      REG              253,0     28652      33619 /data/data/com.example.android.dev/cache/dd-logs-v2/1663001682074 (deleted)
12391    u0_a120  185w      REG              253,0     28652      33619 /data/data/com.example.android.dev/cache/dd-logs-v2/1663001682074 (deleted)
12391    u0_a120  188w      REG              253,0     28652      33619 /data/data/com.example.android.dev/cache/dd-logs-v2/1663001682074 (deleted)
12391    u0_a120  190w      REG              253,0     28652      33619 /data/data/com.example.android.dev/cache/dd-logs-v2/1663001682074 (deleted)
12391    u0_a120  191w      REG              253,0     28652      33619 /data/data/com.example.android.dev/cache/dd-logs-v2/1663001682074 (deleted)
12391    u0_a120  192w      REG              253,0     28652      33619 /data/data/com.example.android.dev/cache/dd-logs-v2/1663001682074 (deleted)
12391    u0_a120  194w      REG              253,0     28652      33619 /data/data/com.example.android.dev/cache/dd-logs-v2/1663001682074 (deleted)
12391    u0_a120  195w      REG              253,0     25403      33647 /data/data/com.example.android.dev/cache/dd-logs-v2/1663001691954
12391    u0_a120  199w      REG              253,0     28652      33619 /data/data/com.example.android.dev/cache/dd-logs-v2/1663001682074 (deleted)
12391    u0_a120  200w      REG              253,0     28652      33619 /data/data/com.example.android.dev/cache/dd-logs-v2/1663001682074 (deleted)
12391    u0_a120  201w      REG              253,0     28652      33619 /data/data/com.example.android.dev/cache/dd-logs-v2/1663001682074 (deleted)
12391    u0_a120  204w      REG              253,0    177722      33682 /data/data/com.example.android.dev/cache/dd-logs-v2/1663001706380
12391    u0_a120  207w      REG              253,0     28652      33619 /data/data/com.example.android.dev/cache/dd-logs-v2/1663001682074 (deleted)
12391    u0_a120  209w      REG              253,0     28652      33619 /data/data/com.example.android.dev/cache/dd-logs-v2/1663001682074 (deleted)
12391    u0_a120  210w      REG              253,0     28652      33619 /data/data/com.example.android.dev/cache/dd-logs-v2/1663001682074 (deleted)
12391    u0_a120  213w      REG              253,0     28652      33619 /data/data/com.example.android.dev/cache/dd-logs-v2/1663001682074 (deleted)
12391    u0_a120  214w      REG              253,0     28652      33619 /data/data/com.example.android.dev/cache/dd-logs-v2/1663001682074 (deleted)
12391    u0_a120  215w      REG              253,0     28652      33619 /data/data/com.example.android.dev/cache/dd-logs-v2/1663001682074 (deleted)
12391    u0_a120  219w      REG              253,0     28652      33619 /data/data/com.example.android.dev/cache/dd-logs-v2/1663001682074 (deleted)
12391    u0_a120  220w      REG              253,0     28652      33619 /data/data/com.example.android.dev/cache/dd-logs-v2/1663001682074 (deleted)
12391    u0_a120  221w      REG              253,0     25403      33647 /data/data/com.example.android.dev/cache/dd-logs-v2/1663001691954
12391    u0_a120  222w      REG              253,0     28652      33619 /data/data/com.example.android.dev/cache/dd-logs-v2/1663001682074 (deleted)
12391    u0_a120  223w      REG              253,0     28652      33619 /data/data/com.example.android.dev/cache/dd-logs-v2/1663001682074 (deleted)
12391    u0_a120  224r      REG              253,0     27916      33607 /data/data/com.example.android.dev/cache/dd-logs-v2/1663001677320 (deleted)
12391    u0_a120  232w      REG              253,0     28716      33632 /data/data/com.example.android.dev/cache/dd-logs-v2/1663001686837 (deleted)
12391    u0_a120  234w      REG              253,0     28652      33619 /data/data/com.example.android.dev/cache/dd-logs-v2/1663001682074 (deleted)
12391    u0_a120  235w      REG              253,0     25403      33647 /data/data/com.example.android.dev/cache/dd-logs-v2/1663001691954
12391    u0_a120  236w      REG              253,0     28652      33619 /data/data/com.example.android.dev/cache/dd-logs-v2/1663001682074 (deleted)
12391    u0_a120  237w      REG              253,0     28652      33619 /data/data/com.example.android.dev/cache/dd-logs-v2/1663001682074 (deleted)
12391    u0_a120  238w      REG              253,0     28652      33619 /data/data/com.example.android.dev/cache/dd-logs-v2/1663001682074 (deleted)
12391    u0_a120  240w      REG              253,0     28652      33619 /data/data/com.example.android.dev/cache/dd-logs-v2/1663001682074 (deleted)
12391    u0_a120  241w      REG              253,0     28652      33619 /data/data/com.example.android.dev/cache/dd-logs-v2/1663001682074 (deleted)
12391    u0_a120  242w      REG              253,0     28652      33619 /data/data/com.example.android.dev/cache/dd-logs-v2/1663001682074 (deleted)
12391    u0_a120  243w      REG              253,0     28652      33619 /data/data/com.example.android.dev/cache/dd-logs-v2/1663001682074 (deleted)
12391    u0_a120  246w      REG              253,0     28652      33619 /data/data/com.example.android.dev/cache/dd-logs-v2/1663001682074 (deleted)
12391    u0_a120  247w      REG              253,0     28652      33619 /data/data/com.example.android.dev/cache/dd-logs-v2/1663001682074 (deleted)
12391    u0_a120  250w      REG              253,0     28652      33619 /data/data/com.example.android.dev/cache/dd-logs-v2/1663001682074 (deleted)
12391    u0_a120  252w      REG              253,0     25403      33647 /data/data/com.example.android.dev/cache/dd-logs-v2/1663001691954
12391    u0_a120  253w      REG              253,0     25403      33647 /data/data/com.example.android.dev/cache/dd-logs-v2/1663001691954
12391    u0_a120  254w      REG              253,0     28716      33632 /data/data/com.example.android.dev/cache/dd-logs-v2/1663001686837 (deleted)
12391    u0_a120  255w      REG              253,0     28716      33632 /data/data/com.example.android.dev/cache/dd-logs-v2/1663001686837 (deleted)
12391    u0_a120  256w      REG              253,0     28716      33632 /data/data/com.example.android.dev/cache/dd-logs-v2/1663001686837 (deleted)
12391    u0_a120  258w      REG              253,0     28716      33632 /data/data/com.example.android.dev/cache/dd-logs-v2/1663001686837 (deleted)
12391    u0_a120  260w      REG              253,0     28716      33632 /data/data/com.example.android.dev/cache/dd-logs-v2/1663001686837 (deleted)
12391    u0_a120  261w      REG              253,0     28716      33632 /data/data/com.example.android.dev/cache/dd-logs-v2/1663001686837 (deleted)
12391    u0_a120  262w      REG              253,0     28716      33632 /data/data/com.example.android.dev/cache/dd-logs-v2/1663001686837 (deleted)
12391    u0_a120  263w      REG              253,0     28716      33632 /data/data/com.example.android.dev/cache/dd-logs-v2/1663001686837 (deleted)
12391    u0_a120  265w      REG              253,0     28716      33632 /data/data/com.example.android.dev/cache/dd-logs-v2/1663001686837 (deleted)
12391    u0_a120  266w      REG              253,0     28716      33632 /data/data/com.example.android.dev/cache/dd-logs-v2/1663001686837 (deleted)
12391    u0_a120  267w      REG              253,0     28716      33632 /data/data/com.example.android.dev/cache/dd-logs-v2/1663001686837 (deleted)
12391    u0_a120  269w      REG              253,0     28716      33632 /data/data/com.example.android.dev/cache/dd-logs-v2/1663001686837 (deleted)
12391    u0_a120  270w      REG              253,0     28716      33632 /data/data/com.example.android.dev/cache/dd-logs-v2/1663001686837 (deleted)
12391    u0_a120  271w      REG              253,0     28716      33632 /data/data/com.example.android.dev/cache/dd-logs-v2/1663001686837 (deleted)
12391    u0_a120  272w      REG              253,0     28716      33632 /data/data/com.example.android.dev/cache/dd-logs-v2/1663001686837 (deleted)
12391    u0_a120  273w      REG              253,0     28716      33632 /data/data/com.example.android.dev/cache/dd-logs-v2/1663001686837 (deleted)
12391    u0_a120  274w      REG              253,0     28716      33632 /data/data/com.example.android.dev/cache/dd-logs-v2/1663001686837 (deleted)
12391    u0_a120  275w      REG              253,0     28716      33632 /data/data/com.example.android.dev/cache/dd-logs-v2/1663001686837 (deleted)
12391    u0_a120  276w      REG              253,0     28716      33632 /data/data/com.example.android.dev/cache/dd-logs-v2/1663001686837 (deleted)
12391    u0_a120  277w      REG              253,0     28716      33632 /data/data/com.example.android.dev/cache/dd-logs-v2/1663001686837 (deleted)
12391    u0_a120  278w      REG              253,0     28716      33632 /data/data/com.example.android.dev/cache/dd-logs-v2/1663001686837 (deleted)
12391    u0_a120  279w      REG              253,0     28716      33632 /data/data/com.example.android.dev/cache/dd-logs-v2/1663001686837 (deleted)
12391    u0_a120  280w      REG              253,0     28716      33632 /data/data/com.example.android.dev/cache/dd-logs-v2/1663001686837 (deleted)
12391    u0_a120  281w      REG              253,0     28716      33632 /data/data/com.example.android.dev/cache/dd-logs-v2/1663001686837 (deleted)
12391    u0_a120  282w      REG              253,0     28716      33632 /data/data/com.example.android.dev/cache/dd-logs-v2/1663001686837 (deleted)
12391    u0_a120  283w      REG              253,0     28716      33632 /data/data/com.example.android.dev/cache/dd-logs-v2/1663001686837 (deleted)
12391    u0_a120  284w      REG              253,0     28716      33632 /data/data/com.example.android.dev/cache/dd-logs-v2/1663001686837 (deleted)
12391    u0_a120  285w      REG              253,0     28716      33632 /data/data/com.example.android.dev/cache/dd-logs-v2/1663001686837 (deleted)
12391    u0_a120  286w      REG              253,0     28716      33632 /data/data/com.example.android.dev/cache/dd-logs-v2/1663001686837 (deleted)
12391    u0_a120  287w      REG              253,0     25403      33647 /data/data/com.example.android.dev/cache/dd-logs-v2/1663001691954
12391    u0_a120  288w      REG              253,0     28716      33632 /data/data/com.example.android.dev/cache/dd-logs-v2/1663001686837 (deleted)
12391    u0_a120  289w      REG              253,0     89375      33672 /data/data/com.example.android.dev/cache/dd-logs-v2/1663001701613
12391    u0_a120  290w      REG              253,0     28716      33632 /data/data/com.example.android.dev/cache/dd-logs-v2/1663001686837 (deleted)
12391    u0_a120  291w      REG              253,0     25403      33647 /data/data/com.example.android.dev/cache/dd-logs-v2/1663001691954
12391    u0_a120  293w      REG              253,0     25403      33647 /data/data/com.example.android.dev/cache/dd-logs-v2/1663001691954

+ 415 dimilar lines removed
```

**Comment by 0xnm** [comment]
Hello @ntoskrnl!

I cannot reproduce this issue on SDK version `1.13.0` + emulator running Android 11. Can you please give us more details?

* If there are any specific repro steps, please share them.
* Can you also give us the SDK configuration?
* Do you have any specific network settings on your emulator? Like unstable/poor network connection, etc.
* What is the minimum time span to start observing the results you have?

**Comment by ntoskrnl** [comment]
Hi! Thanks for quick reply. Looks like it is not an issue in DataDog android sdk.

My app doesn't create file descriptors and only after adding the DataDog sdk I saw this behaviour. Naturally I thought it might be related to the SDK.

After doing the same profiling with the sample app, I don't see any leaking file descriptors. Moreover, I wrote a simple test with a loop, where I write to files, close and delete them. I saw 0 file descriptors leaked on the sample app, but on my app all of them leaked. 

This is definitely not an issue with the SDK, probably some configuration/dependency on my app. I have no idea yet what can cause such behaviour though, but it prevents me from enabling DD sdk in the app, will investigate it later.

Closing this ticket now.
